src/scene/Texture.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 Robotec.AI
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <scene/Texture.hpp>
+#include <cuda_runtime.h>
+
+API_OBJECT_INSTANCE(Texture);
+
+Texture::Texture(const void* texels, int width, int height) try :
+		resolution(width, height)
+		{
+			createTextureObject(texels, width, height);
+		}
+		catch (const std::exception& e)
+		{
+			cleanup();
+		}
+
+void Texture::createTextureObject(const void* texels, int width, int height)
+{
+	cudaResourceDesc res_desc = {};
+
+	int32_t numComponents = 1;
+
+	cudaChannelFormatDesc channel_desc = cudaCreateChannelDesc<uchar1>();
+
+	int32_t pitch = width * numComponents * sizeof(unsigned char);
```

Comment:
I don't like that the type of texels is repeated in multiple places (here, `optixPrograms.cu`, `apiCore.cpp`, in tests). I'd define a separate type for it. It has already caused a bug pointed out in this comment.